src/network/address.rs

```diff
@@ -33,6 +34,32 @@ pub struct Address {
     pub port: u16
 }
 
+impl Address {
+    /// Create an address message for a socket
+    pub fn new (socket :&SocketAddr, services: u64) -> Address {
+        let (address, port) = match socket {
+            &SocketAddr::V4(ref addr) => (addr.ip().to_ipv6_mapped().segments(), addr.port()),
+            &SocketAddr::V6(ref addr) => (addr.ip().segments(), addr.port())
+        };
+        Address { address, port, services }
+    }
+
+    /// extract socket address from an address message
+    pub fn socket_addr (&self) -> SocketAddr {
```

Comment:
You should have a check that the contained addr isnt a Tor-wrapped one and return a Result<>, sadly, I think.

Reply:
Please give me a pointer how that is recognized. I did not yet deal with tor at this level.